### PR TITLE
Update manager when switching workspace root in a multi-root workspace

### LINKED ISSUE #3162: File watcher and structure are wrong when changing rootFile in a multi-root workspace
Consider the multi-root workspace

```
event.code-workspace

A/
  .vscode/settings.json
  A.tex

B/
  B.tex
```

1. Visit the two files `A.text` and `B.tex` once. This is important to make sure the extension knows about these two files.
2. Changing from `A.tex` to `B.tex` and vice versa does not update the outline or structure view from the TeX badge. Moreover, the list of watched files remains unchanged.

We do not enter in https://github.com/James-Yu/LaTeX-Workshop/blob/master/src/components/manager.ts#L343-L355 because  `A.tex` is the rootFile associated to workspace `A` and `B.tex` is the rootFile associated to workspace `B`. Note that `Manager.rootFile` returns the rootFile for the current workspace.

Here are the logs

```
[22:54:15] Initializing LaTeX Workshop.
[22:54:15] Extension root: /Users/jl/devel/LaTeX-Workshop
[22:54:15] $PATH: /Users/jl/.nvm/versions/node/v17.3.0/bin:/usr/local/opt/python/libexec/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/Users/jl/bin:/Users/jl/live/teaching/ensimag/tools/chamilotools
[22:54:15] $SHELL: /bin/zsh
[22:54:15] $LANG: en_US.UTF-8
[22:54:15] $LC_ALL: undefined
[22:54:15] process.platform: darwin
[22:54:15] process.arch: x64
[22:54:15] vscode.env.appName: Visual Studio Code
[22:54:15] vscode.env.remoteName: undefined
[22:54:15] vscode.env.uiKind: 1
[22:54:15] Workspace for configuration: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A
[22:54:15] editor.acceptSuggestionOnEnter: "off"
[22:54:15] latex-workshop.bind.enter.key: true
[22:54:15] latex-workshop.docker.enabled: false
[22:54:15] latex-workshop.docker.image.latex: ""
[22:54:15] latex-workshop.hover.preview.mathjax.extensions: []
[22:54:15] latex-workshop.intellisense.package.enabled: true
[22:54:15] latex-workshop.intellisense.update.aggressive.enabled: false
[22:54:15] latex-workshop.intellisense.update.delay: 1000
[22:54:15] latex-workshop.latex.autoBuild.run: "never"
[22:54:15] latex-workshop.latex.build.forceRecipeUsage: true
[22:54:15] latex-workshop.latex.outDir: "%DIR%"
[22:54:15] latex-workshop.latex.recipes: [
 {
  "name": "latexmk 🔃",
  "tools": [
   "latexmk"
  ]
 }
]
[22:54:15] latex-workshop.latex.tools: [
 {
  "name": "latexmk",
  "command": "latexmk",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "-pdf",
   "-outdir=%OUTDIR%",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "lualatexmk",
  "command": "latexmk",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "-lualatex",
   "-outdir=%OUTDIR%",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "latexmk_rconly",
  "command": "latexmk",
  "args": [
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "pdflatex",
  "command": "pdflatex",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "bibtex",
  "command": "bibtex",
  "args": [
   "%DOCFILE%"
  ],
  "env": {}
 },
 {
  "name": "rnw2tex",
  "command": "Rscript",
  "args": [
   "-e",
   "knitr::opts_knit$set(concordance = TRUE); knitr::knit('%DOCFILE_EXT%')"
  ],
  "env": {}
 },
 {
  "name": "jnw2tex",
  "command": "julia",
  "args": [
   "-e",
   "using Weave; weave(\"%DOC_EXT%\", doctype=\"tex\")"
  ],
  "env": {}
 },
 {
  "name": "jnw2texmintex",
  "command": "julia",
  "args": [
   "-e",
   "using Weave; weave(\"%DOC_EXT%\", doctype=\"texminted\")"
  ],
  "env": {}
 },
 {
  "name": "tectonic",
  "command": "tectonic",
  "args": [
   "--synctex",
   "--keep-logs",
   "%DOC%.tex"
  ],
  "env": {}
 }
]
[22:54:15] latex-workshop.viewer.pdf.internal.keyboardEvent: "auto"
[22:54:15] Workspace for configuration: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B
[22:54:15] editor.acceptSuggestionOnEnter: "off"
[22:54:15] latex-workshop.bind.enter.key: true
[22:54:15] latex-workshop.docker.enabled: false
[22:54:15] latex-workshop.docker.image.latex: ""
[22:54:15] latex-workshop.hover.preview.mathjax.extensions: []
[22:54:15] latex-workshop.intellisense.package.enabled: true
[22:54:15] latex-workshop.intellisense.update.aggressive.enabled: false
[22:54:15] latex-workshop.intellisense.update.delay: 1000
[22:54:15] latex-workshop.latex.autoBuild.run: "never"
[22:54:15] latex-workshop.latex.build.forceRecipeUsage: true
[22:54:15] latex-workshop.latex.outDir: "%DIR%"
[22:54:15] latex-workshop.latex.recipes: [
 {
  "name": "latexmk 🔃",
  "tools": [
   "latexmk"
  ]
 },
 {
  "name": "latexmk (latexmkrc)",
  "tools": [
   "latexmk_rconly"
  ]
 },
 {
  "name": "latexmk (lualatex)",
  "tools": [
   "lualatexmk"
  ]
 },
 {
  "name": "pdflatex ➞ bibtex ➞ pdflatex × 2",
  "tools": [
   "pdflatex",
   "bibtex",
   "pdflatex",
   "pdflatex"
  ]
 },
 {
  "name": "Compile Rnw files",
  "tools": [
   "rnw2tex",
   "latexmk"
  ]
 },
 {
  "name": "Compile Jnw files",
  "tools": [
   "jnw2tex",
   "latexmk"
  ]
 },
 {
  "name": "tectonic",
  "tools": [
   "tectonic"
  ]
 }
]
[22:54:15] latex-workshop.latex.tools: [
 {
  "name": "latexmk",
  "command": "latexmk",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "-pdf",
   "-outdir=%OUTDIR%",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "lualatexmk",
  "command": "latexmk",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "-lualatex",
   "-outdir=%OUTDIR%",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "latexmk_rconly",
  "command": "latexmk",
  "args": [
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "pdflatex",
  "command": "pdflatex",
  "args": [
   "-synctex=1",
   "-interaction=nonstopmode",
   "-file-line-error",
   "%DOC%"
  ],
  "env": {}
 },
 {
  "name": "bibtex",
  "command": "bibtex",
  "args": [
   "%DOCFILE%"
  ],
  "env": {}
 },
 {
  "name": "rnw2tex",
  "command": "Rscript",
  "args": [
   "-e",
   "knitr::opts_knit$set(concordance = TRUE); knitr::knit('%DOCFILE_EXT%')"
  ],
  "env": {}
 },
 {
  "name": "jnw2tex",
  "command": "julia",
  "args": [
   "-e",
   "using Weave; weave(\"%DOC_EXT%\", doctype=\"tex\")"
  ],
  "env": {}
 },
 {
  "name": "jnw2texmintex",
  "command": "julia",
  "args": [
   "-e",
   "using Weave; weave(\"%DOC_EXT%\", doctype=\"texminted\")"
  ],
  "env": {}
 },
 {
  "name": "tectonic",
  "command": "tectonic",
  "args": [
   "--synctex",
   "--keep-logs",
   "%DOC%.tex"
  ],
  "env": {}
 }
]
[22:54:15] latex-workshop.viewer.pdf.internal.keyboardEvent: "auto"
[22:54:15] Creating a new file watcher.
[22:54:15] watcherOptions: {"useFsEvents":false,"usePolling":false,"interval":300,"binaryInterval":1000,"awaitWriteFinish":{"stabilityThreshold":250}}
[22:54:15] Creating PDF file watcher.
[22:54:15] watcherOptions: {"useFsEvents":false,"usePolling":false,"interval":300,"binaryInterval":1000,"awaitWriteFinish":{}}
[22:54:15] Creating Bib file watcher.
[22:54:15] watcherOptions: {"useFsEvents":false,"usePolling":false,"interval":300,"binaryInterval":1000,"awaitWriteFinish":{"stabilityThreshold":250}}
[22:54:15] Set $LATEXWORKSHOP_DOCKER_LATEX: ""
[22:54:15] [Server] Creating LaTeX Workshop http and websocket server.
[22:54:15] Bibtex format config: {"tab":"  ","case":"lowercase","left":"{","right":"}","trailingComma":false,"sort":["key"],"alignOnEqual":true,"sortFields":false,"fieldsOrder":[],"firstEntries":["string","xdata"]}
[22:54:15] LaTeX Workshop initialized.
[22:54:15] Bibtex format config: {"tab":"  ","case":"lowercase","left":"{","right":"}","trailingComma":false,"sort":["key"],"alignOnEqual":true,"sortFields":false,"fieldsOrder":[],"firstEntries":["string","xdata"]}
[22:54:15] Trigger characters for intellisense of LaTeX documents: ["\\",",","{"]
[22:54:15] Bibtex format config: {"tab":"  ","case":"lowercase","left":"{","right":"}","trailingComma":false,"sort":["key"],"alignOnEqual":true,"sortFields":false,"fieldsOrder":[],"firstEntries":["string","xdata"]}
[22:54:15] Current workspace folders: ["file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A","file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B"]
[22:54:15] Current workspaceRootDir: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B
[22:54:15] Found root file from active editor: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:15] Root file changed: from undefined to /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:15] Start to find all dependencies.
[22:54:15] Root file languageId: latex
[22:54:15] [Server] Server successfully started: {"address":"127.0.0.1","family":"IPv4","port":60282}
[22:54:15] Reset file watcher.
[22:54:15] Parsing a file and its subfiles: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:15] Parse fls file.
[22:54:15] Fls file found: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.fls
[22:54:15] Parse aux file: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.aux
[22:54:21] Linter for root file started.
[22:54:21] Linter for root file command: chktex
[22:54:21] Linter for root file command args: ["-wall","-e16","-n1","-n3","-n6","-n12","-n19","-n21","-n22","-n24","-n25","-n30","-n36","-n37","-n44","-n46","-q","-f%f:%l:%c:%d:%k:%n:%m\n","/Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex"]
[22:54:21] Added to file watcher: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:21] Snippet data loaded.
[22:54:21] Checking for duplicate labels: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex.
[22:54:21] [Server] valdOrigin is http://127.0.0.1:60282
[22:54:21] Linter for root file successfully finished in 0s 354ms
[22:54:21] The .chktexrc file not found.
[22:54:21] Linter log parsed with 5 messages.
[22:54:23] Manager.fileWatcher.getWatched: {"/Users/jl/devel/LaTeX-Workshop/mytests/3082/B":["B.tex"]}
[22:54:23] Manager.filesWatched: ["/Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex"]
[22:54:23] BibWatcher.bibWatcher.getWatched: {}
[22:54:23] BibWatcher.bibsWatched: []
[22:54:23] PdfWatcher.pdfWatcher.getWatched: {}
[22:54:23] PdfWatcher.pdfsWatched: []
[22:54:23] PdfWatcher.watchedPdfVirtualUris: []
[22:54:23] PdfWatcher.ignoredPdfUris: []
[22:54:31] Current workspace folders: ["file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A","file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B"]
[22:54:31] Current workspaceRootDir: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A
[22:54:31] Current workspace folders: ["file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A","file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B"]
[22:54:31] Current workspaceRootDir: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A
[22:54:31] Found root file from active editor: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:31] Found root file from active editor: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:31] Root file changed: from undefined to /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:31] Start to find all dependencies.
[22:54:31] Root file languageId: latex
[22:54:31] Keep using the same root file: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:31] Reset file watcher.
[22:54:31] Parsing a file and its subfiles: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:31] Parse fls file.
[22:54:31] Fls file found: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.fls
[22:54:31] Parse aux file: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.aux
[22:54:31] Added to file watcher: /Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex
[22:54:33] Manager.fileWatcher.getWatched: {"/Users/jl/devel/LaTeX-Workshop/mytests/3082/A":["A.tex"]}
[22:54:33] Manager.filesWatched: ["/Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex"]
[22:54:33] BibWatcher.bibWatcher.getWatched: {}
[22:54:33] BibWatcher.bibsWatched: []
[22:54:33] PdfWatcher.pdfWatcher.getWatched: {}
[22:54:33] PdfWatcher.pdfsWatched: []
[22:54:33] PdfWatcher.watchedPdfVirtualUris: []
[22:54:33] PdfWatcher.ignoredPdfUris: []
[22:54:33] Manager.fileWatcher.getWatched: {"/Users/jl/devel/LaTeX-Workshop/mytests/3082/A":["A.tex"]}
[22:54:33] Manager.filesWatched: ["/Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex"]
[22:54:33] BibWatcher.bibWatcher.getWatched: {}
[22:54:33] BibWatcher.bibsWatched: []
[22:54:33] PdfWatcher.pdfWatcher.getWatched: {}
[22:54:33] PdfWatcher.pdfsWatched: []
[22:54:33] PdfWatcher.watchedPdfVirtualUris: []
[22:54:33] PdfWatcher.ignoredPdfUris: []
[22:54:57] Current workspace folders: ["file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/A","file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B"]
[22:54:57] Current workspaceRootDir: file:///Users/jl/devel/LaTeX-Workshop/mytests/3082/B
[22:54:57] Found root file from active editor: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:57] Keep using the same root file: /Users/jl/devel/LaTeX-Workshop/mytests/3082/B/B.tex
[22:54:57] Linter log parsed with 5 messages.
[22:54:59] Manager.fileWatcher.getWatched: {"/Users/jl/devel/LaTeX-Workshop/mytests/3082/A":["A.tex"]}
[22:54:59] Manager.filesWatched: ["/Users/jl/devel/LaTeX-Workshop/mytests/3082/A/A.tex"]
```

### PR DESCRIPTION
**Pull request by jlelong**
Close #3162

### CODE REVIEW

**Comment by tamuratak** [comment]
We had better stop using `Manager.rootFiles`.  See

- https://github.com/tamuratak/LaTeX-Workshop/commit/c7af08d53028b7076fa62a14b755f3b4e05acf59

There is no reason to cache rootfiles. And, caches could cause problems we are seeing.

**Comment by tamuratak** [comment]
We could say the same thing with `Manager.localRootFiles`.

**Comment by jlelong** [comment]
> There is no reason to cache rootfiles. And, caches could cause problems we are seeing.

I have to investigate this more closely but I was actually wondering the point of caching the rootFile for every workspace. This would make the `Manager` simpler.

File: src/components/manager.ts [hunk 1]
```diff
@@ -84,10 +84,9 @@ export class Manager {
      */
     private readonly cachedContent = Object.create(null) as Content
 
-    private readonly localRootFiles = Object.create(null) as { [key: string]: string | undefined }
-    private readonly rootFilesLanguageIds = Object.create(null) as { [key: string]: string | undefined }
-    // Store one root file for each workspace.
-    private readonly rootFiles = Object.create(null) as { [key: string]: RootFileType | undefined }
+    private _localRootFile: string | undefined
+    private _rootFilesLanguageId: string | undefined
```
Comment (tamuratak) [inline src/components/manager.ts:88]: Singular form `_rootFileLanguageId` would be better.

**Review by tamuratak — CHANGES_REQUESTED** [review]

File: src/components/manager.ts (hunk 1 above)
Comment (jlelong) [reply src/components/manager.ts]: For sure!

**Comment by tamuratak** [comment]
Do we really need to store `Manager.workspaceRootDirUri`? We can use `vscode.workspace.getWorkspaceFolder(manager.rootFileUri)` to obtain the workspace folder that includes a rootfile.

In the current implementation, `Manager.workspaceRootDirUri` and the workspace folder that includes a rootfile can be different. It would make things complicated. We had better use `Manager.findWorkspace` only for `Manager.findRootInWorkspace`

https://github.com/James-Yu/LaTeX-Workshop/blob/73133df3f3c172b5ab3d897bb79bec3036f48513/src/components/manager.ts#L81-L91

**Comment by jlelong** [comment]
You are right, `Manager.workspaceRootDirUri` is not needed anymore. I have just removed it.

**Review by tamuratak — APPROVED** [review]

**Comment by jlelong** [comment]
Btw, the test for the structure provider always fails under Windows https://github.com/jlelong/LaTeX-Workshop/runs/5431416245?check_suite_focus=true

Increasing the waiting time for the extension to update to  `6000` does not help. I have no idea what is causing this. Maybe, we should disable this test under Windows?

**Comment by tamuratak** [comment]
> Maybe, we should disable this test under Windows?

Yes. Some tests fail. I have not been able to find reasons.

**Pull request merged**